src/main/java/jenkins/branch/WorkspaceLocatorImpl.java

```diff
@@ -438,6 +450,51 @@ private static synchronized void taskFinished() {
             Deleter.class.notifyAll();
         }
 
+        private static class CleanupTaskProvisioner implements Runnable{
+
+            @NonNull
+            private final TopLevelItem tli;
+
+            @NonNull
+            private final Queue<Node> nodes;
+
+            private final int threadLimit;
+
+            public CleanupTaskProvisioner(TopLevelItem tli, List<Node> nodes, int threadLimit) {
+                this.tli = tli;
+                this.nodes = new LinkedList<>(nodes);
+                this.threadLimit = threadLimit;
+                if (threadLimit > 0) {
+                    LOGGER.log(Level.INFO, "Cleanup thread limit set: {0}", threadLimit);
+                }
+            }
+
+            @Override
+            public void run() {
+                try {
+                    while (!nodes.isEmpty()){
+                        if (hasFreeThreadVolume(threadLimit)) {
+                            Computer.threadPoolForRemoting.submit(new CleanupTask(tli, nodes.remove()));
+                        } else {
+                            Thread.sleep(1000);
+                        }
+                        LOGGER.log(Level.INFO, "LIVE THREAD COUNT : {0}", ManagementFactory.getThreadMXBean().getThreadCount());
+                    }
+                } catch (Exception e) {
+                    LOGGER.log(Level.WARNING, e.getMessage());
+                }
+            }
+
+            private boolean hasFreeThreadVolume(int threadLimit){
+                if (threadLimit <= 0) {
+                    return true;
+                }
+                int currentThreadCount = ManagementFactory.getThreadMXBean().getThreadCount();
```

Comment:
These are good improvements. But can it be pushed into an improvement work package? Just to get out of the current situation first?

I don't think most users will set this System property. And the ones who are experiencing the repeated server crashes like us, just want it somewhere within the max nproc limit of the user running Jenkins. It is not important for us that cleanup is done through the branch api plugin specifically since we manage that ourselves anyway (and we have many other jobs which aren't using the branch-api cleanup feature to keep disk space usage low)

Reply:
Hey James, thanks a lot for the feedback. It definitely seems reasonable. I tidied up the property parsing, and gave using the Semaphore a shot.

Originally, I ended up passing the Semaphore down to the `CleanupTaskProvider`/`CleanupTask`. I ended up having to check if there even was a limit before trying to `acquire`/`release` in multiple places, which made it all a bit messy. I believe it would work if we had a (large?) limit by default, then no checks would've been required.

Ultimately, I opted for adding two static methods to `Deleter`. Not sure if I like it, but it does make it easier to use.
Thoughts?

Reply:
Thanks @praqma-thi 
@jtnord what are your thoughts on the current state of this PR?

Reply:
Hi @praqma-thi 

> Ultimately, I opted for adding two static methods to Deleter. Not sure if I like it, but it does make it easier to use.
Thoughts?

the Listener is a singleton `Extension` so you can obtain it at runtime with `ExtensionList.lookupSingleton()` to avoid having to have something static or pass the code around.

if you want to try that I will hold merging for a short while. otherwise let me know to merge release

Reply:
Hey James, happy to give that a shot when I'm back from my holiday, plus tidying up the exception handling mentioned below. It'll be next year for me, but I'll gladly dig into it regardless of whether it gets merged now or not.

Reply:
Hi James, I finally got out of hibernation. I rebased onto master and replaced the static methods with getting a reference using `lookupSingleton()`. I'm quite happy with the changes, if there's more improvements, feel free to let me know.